src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/Reliability/SqlRetryPolicy.cs

```diff
@@ -0,0 +1,241 @@
+        public SqlRetryPolicy(ITransientErrorDetectionStrategy errorDetectionStrategy, int retryCount, TimeSpan initialInterval, TimeSpan increment)
+            : this(errorDetectionStrategy, new Incremental(retryCount, initialInterval, increment))
+        {
+        }
+
+        /// <summary>
+        /// An instance of a callback delegate that will be invoked whenever a retry condition is encountered.
+        /// </summary>
+        public event EventHandler<SqlRetryingEventArgs> Retrying;
+
+        /// <summary>
+        /// Gets the retry strategy.
+        /// </summary>
+        public SqlRetryStrategy RetryStrategy { get; private set; }
+
+        /// <summary>
+        /// Gets the instance of the error detection strategy.
+        /// </summary>
+        public ITransientErrorDetectionStrategy ErrorDetectionStrategy { get; private set; }
+
+        /// <summary>
+        /// Repetitively executes the specified action while it satisfies the current retry policy.
+        /// </summary>
+        /// <param name="action">A delegate that represents the executable action that doesn't return any results.</param>
+        public virtual void ExecuteAction(Action action)
+        {
+            this.ExecuteAction(() => { action(); return default(object); });
+        }
+
+        /// <summary>
+        /// Repetitively executes the specified action while it satisfies the current retry policy.
+        /// </summary>
+        /// <typeparam name="TResult">The type of result expected from the executable action.</typeparam>
+        /// <param name="func">A delegate that represents the executable action that returns the result of type <typeparamref name="TResult"/>.</param>
+        /// <returns>The result from the action.</returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", MessageId = "0", Justification = "Validated with Guard")]
+        public virtual TResult ExecuteAction<TResult>(Func<TResult> func)
+        {
+            int retryCount = 0;
+            TimeSpan delay = TimeSpan.Zero;
+            Exception lastError;
+
+            var shouldRetry = this.RetryStrategy.GetShouldRetry();
+
+            for (;;)
+            {
+                lastError = null;
+
+                try
+                {
+                    return func();
+                }
+                catch (Exception ex)
+                {
+                    lastError = ex;
+
+                    if (!(this.ErrorDetectionStrategy.IsTransient(lastError) && shouldRetry(retryCount++, lastError, out delay)))
+                    {
+                        throw;
+                    }
+                }
+
+                // Perform an extra check in the delay interval. Should prevent from accidentally ending up with the value of -1 that will block a thread indefinitely. 
```

Comment:
Invalid delays such as negative values (e.g. minus 42 seconds) should throw an exception. These are logic bugs. They should not be silenced.